plugins/experimental/certifier/certifier.cc

```diff
@@ -0,0 +1,691 @@
+    TSError("[%s] mkcsr(): Failed to add entry.", PLUGIN_NAME);
+    return nullptr;
+  }
+  /// Set Traffic Server public key
+  if(X509_REQ_set_pubkey(req.get(), ca_pkey_scoped.get())!=1) {
+    TSError("[%s] mkcsr(): Failed to set pubkey.", PLUGIN_NAME);
+    return nullptr;
+  }
+  /// Sign with Traffic Server private key
+  if(X509_REQ_sign(req.get(), ca_pkey_scoped.get(),EVP_sha256())<=0) {
+    TSError("[%s] mkcsr(): Failed to Sign.", PLUGIN_NAME);
+    return nullptr;
+  }
+  return req;
+}
+
+/// Local helper function that signs a X509 certificate with (CA) private key
+static int
+sign_X509(X509 *cert, EVP_PKEY *pkey, const EVP_MD *md)
+{
+  int rv;
+  EVP_MD_CTX mctx;
+  EVP_PKEY_CTX *pkctx = NULL;
+
+  /// Init message digest context
+  EVP_MD_CTX_init(&mctx);
+  rv = EVP_DigestSignInit(&mctx, &pkctx, md, NULL, pkey);
+
+  /// Sign X509
+  if (rv > 0) {
+    rv = X509_sign_ctx(cert, &mctx);
+  }
+
+  EVP_MD_CTX_cleanup(&mctx);
+  return rv > 0 ? 1 : 0;
+}
+
+/// Local helper function that generates a X509 certificate based on CSR
+static scoped_X509
+mkcrt(X509_REQ *req, int serial)
+{
+  TSDebug(PLUGIN_NAME, "Entering mkcrt()...");
+  X509_NAME * subj, *tmpsubj;
+  scoped_EVP_PKEY pktmp;
+  scoped_X509 cert;
+
+  cert.reset(X509_new());
+
+  /// Set X509V3
+  if (X509_set_version(cert.get(), 2) == 0) {
+    TSError("[%s] mkcrt(): Failed to set X509V3.", PLUGIN_NAME);
+    return nullptr;
+  }
+
+  /// Set serial number
+  //TSDebug("txn_monitor", "serial: %d", serial);
+  ASN1_INTEGER_set(X509_get_serialNumber(cert.get()), serial);
+
+  /// Set issuer from CA cert
+  if (X509_set_issuer_name(cert.get(), X509_get_subject_name(ca_cert_scoped.get())) == 0) {
+    TSError("[%s] mkcrt(): Failed to set issuer.", PLUGIN_NAME);
+    return nullptr;
+  }
+  /// Set certificate time
+  X509_gmtime_adj(X509_get_notBefore(cert.get()), 0);
+  X509_gmtime_adj(X509_get_notAfter(cert.get()), (long)3650*24*3600);
+
+  /// Get a handle to csr subject name
+  subj = X509_REQ_get_subject_name(req);
+  if ((tmpsubj = X509_NAME_dup(subj)) == nullptr) {
+    TSDebug(PLUGIN_NAME,"mkcrt(): Failed to duplicate subject name.");
+    return nullptr;
+  }
+  if ((X509_set_subject_name(cert.get(), tmpsubj)) == 0) {
+    TSDebug(PLUGIN_NAME, "mkcrt(): Failed to set X509 subject name");
+    X509_NAME_free(tmpsubj);    ///< explicit call to free X509_NAME object
+    return nullptr;
+  }
+  pktmp.reset(X509_REQ_get_pubkey(req));
+  if (pktmp == nullptr) {
+    TSDebug(PLUGIN_NAME, "mkcrt(): Failed to get CSR public key.");
+    X509_NAME_free(tmpsubj);
+    return nullptr;
+  }
+  if (X509_set_pubkey(cert.get(), pktmp.get()) == 0) {
+    TSDebug(PLUGIN_NAME, "mkcrt(): Failed to set X509 public key.");
+    X509_NAME_free(tmpsubj);
+    return nullptr;
+  }
+
+  sign_X509(cert.get(), ca_pkey_scoped.get(), EVP_sha256());
+
+  return cert;
+}
+
+static int
+shadow_cert_generator(TSCont contp, TSEvent event, void *edata)
+{
```

Comment:
I suggest using the Continuation class in the recently-added Continuation.h header file.